File: tests-python/pages/register_page.py
from __future__ import annotations

import allure
from selenium.webdriver.common.by import By

from pages.base import BasePage


class RegisterPage(BasePage):
    LOGIN_INPUT = (By.CSS_SELECTOR, "[data-testid='login-input']")
    PASSWORD_INPUT = (By.CSS_SELECTOR, "[data-testid='password-input']")
    CONFIRM_PASSWORD = (By.CSS_SELECTOR, "[data-testid='confirm-password-input']")
    SUBMIT = (By.CSS_SELECTOR, "[data-testid='submit-button']")
    FORM_TITLE = (By.CSS_SELECTOR, "[data-testid='register-form-title']")
    ERROR = (By.CSS_SELECTOR, "[data-testid='error-message']")
    LOGIN_LINK = (By.CSS_SELECTOR, "[data-testid='login-link']")
    REGISTER_FORM = (By.CSS_SELECTOR, "[data-testid='register-form']")

    @allure.step("Open register page")
    def open_page(self) -> RegisterPage:
        self.open_path("/register")
        return self

    @allure.step("Click 'Login' link under the register form")
    def click_login_link(self):
        from pages.login_page import LoginPage

        self.find(*self.LOGIN_LINK).click()
        return LoginPage(self.driver, self.config)

    @allure.step("Fill and submit register form")
    def fill_and_submit_form(self, username: str, password: str, confirm_password: str):
        self.type_username(username)
        self.type_password(password)
        self.type_confirm_password(confirm_password)
        return self.submit()

    @allure.step("Type username: {username}")
    def type_username(self, username: str) -> RegisterPage:
        self.fill(*self.LOGIN_INPUT, username)
        return self

    @allure.step("Type password")
    def type_password(self, password: str) -> RegisterPage:
        self.fill(*self.PASSWORD_INPUT, password)
        return self

    @allure.step("Type confirm password")
    def type_confirm_password(self, confirm_password: str) -> RegisterPage:
        self.fill(*self.CONFIRM_PASSWORD, confirm_password)
        return self

    @allure.step("Submit register form")
    def submit(self):
        from pages.home_page import HomePage

        self.find(*self.SUBMIT).click()
        self.wait_url_is_home()
        return HomePage(self.driver, self.config)

    @allure.step("Verify register form is mounted")
    def should_show_register_form(self) -> RegisterPage:
        self.wait_visible(*self.FORM_TITLE)
        self.wait_visible(*self.LOGIN_INPUT)
        self.wait_visible(*self.PASSWORD_INPUT)
        self.wait_visible(*self.CONFIRM_PASSWORD)
        self.wait_visible(*self.SUBMIT)
        return self

File: java/gradle-junit5-selenium/src/test/java/pages/RegisterPage.java
package pages;

import config.TestConfig;
import io.qameta.allure.Step;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class RegisterPage extends BasePage {

    private static final By LOGIN_INPUT = By.cssSelector("[data-testid='login-input']");
    private static final By PASSWORD_INPUT = By.cssSelector("[data-testid='password-input']");
    private static final By CONFIRM_PASSWORD = By.cssSelector("[data-testid='confirm-password-input']");
    private static final By SUBMIT = By.cssSelector("[data-testid='submit-button']");
    private static final By FORM_TITLE = By.cssSelector("[data-testid='register-form-title']");
    private static final By ERROR = By.cssSelector("[data-testid='error-message']");
    private static final By LOGIN_LINK = By.cssSelector("[data-testid='login-link']");

    public RegisterPage(WebDriver driver, TestConfig config) {
        super(driver, config);
    }

    @Step("Open register page")
    public RegisterPage openPage() {
        openPath("/register");
        return this;
    }

    @Step("Click 'Login' link under the register form")
    public LoginPage clickLoginLink() {
        driver.findElement(LOGIN_LINK).click();
        return new LoginPage(driver, config);
    }

    @Step("Fill and submit register form")
    public HomePage fillAndSubmitForm(String username, String password, String confirmPassword) {
        typeUsername(username);
        typePassword(password);
        typeConfirmPassword(confirmPassword);
        return submit();
    }

    @Step("Type username: {username}")
    public RegisterPage typeUsername(String username) {
        fill(LOGIN_INPUT, username);
        return this;
    }

    @Step("Type password")
    public RegisterPage typePassword(String password) {
        fill(PASSWORD_INPUT, password);
        return this;
    }

    @Step("Type confirm password")
    public RegisterPage typeConfirmPassword(String confirmPassword) {
        fill(CONFIRM_PASSWORD, confirmPassword);
        return this;
    }

    @Step("Submit register form")
    public HomePage submit() {
        driver.findElement(SUBMIT).click();
        waitUrlIsHome();
        return new HomePage(driver, config);
    }

    @Step("Verify register form is mounted")
    public RegisterPage shouldShowRegisterForm() {
        waitVisible(FORM_TITLE);
        waitVisible(LOGIN_INPUT);
        waitVisible(PASSWORD_INPUT);
        waitVisible(CONFIRM_PASSWORD);
        waitVisible(SUBMIT);
        return this;
    }

    @Step("Verify form title message: {message}")
    public RegisterPage shouldHaveFormTitle(String message) {
        waitTextContains(FORM_TITLE, message);
        return this;
    }

    @Step("Verify error message: {message}")
    public RegisterPage shouldHaveErrorMessage(String message) {
        waitVisible(ERROR);
        waitTextContains(ERROR, message);
        return this;
    }
}
